### PR TITLE
Introduce set with custom equals and getHashCode

### PR DESCRIPTION
**Pull request by amcasey**
I got tired of not having custom hash sets in JS so I made a naive one.

Not using linear search to de-dup rename locations cuts search for locations of `Type` from 1750 ms to 1450 ms.

I plan to use this in a bunch of other places in my FAR cleanup.

TODO: this is the simplest possible implementation
TODO: these were the consumers that I'd encountered recently

### CODE REVIEW

File: src/compiler/core.ts [hunk 1]
```diff
@@ -1488,6 +1488,119 @@ namespace ts {
         return createMultiMap() as UnderscoreEscapedMultiMap<T>;
     }
 
+    export function createSet<TElement, THash = number>(getHashCode: (element: TElement) => THash, equals: EqualityComparer<TElement>): Set<TElement> {
```
Comment (sandersn) [inline src/compiler/core.ts:1491]: if this stays in `src/compiler`, I think it should have a longer name to distinguish it from the built-in set, something that reflects that you can customise the equality predicate.

**Review by sandersn — APPROVED** [review]
This seems like a reasonable thing to have, although I'm tempted to say it should be in src/server since it's used only there. Probably having it in a central place is more valuable though.

**Comment by andrewbranch** [comment]
Do you have a sense of how often you expect entries in the set to share a hash? If best practice is for hashes to be unique whenever possible (and it may not be, I don’t know), I wonder if it would be better to let the underlying map values be `TElement | TElement[]`, promoting to an array only once two entries share a hash. That could make for a lot fewer array allocations for large sets with unique hashes, at the expense of probing the array-likeness of the values, lack of support for `TElement` itself being array-like, and polymorphism when some entries do share a hash. FWIW I tried this with the `ExportInfoMap` which is a multimap where you have multiple values when a symbol is re-exported, so in practice that meant that a majority of values were singles but arrays were not uncommon. I couldn’t measure any difference between the always-arrays and arrays-when-needed approaches, so I went with the simpler always-arrays approach like you have here. But I’m curious if you’ve thought about or tried the same thing.

**Comment by amcasey** [comment]
@andrewbranch Generally speaking, I'd expect collisions to be pretty common in a hashset, but I have no particular objection to dynamically promoting values to arrays - this is just the simplest code that solved my problem.

Any thoughts on @sandersn's suggestion of clearly separating this from `Set<T>`?

**Comment by andrewbranch** [comment]
> Any thoughts on @sandersn's suggestion of clearly separating this from `Set<T>`?

Personally, I don’t think it’s much of an issue since `new Set()` ought to be the obvious default to reach for when creating a Set, and the `getHashCode` and `equals` parameters are required in your API.

**Review by andrewbranch — APPROVED** [review]

File: src/compiler/core.ts (hunk 1 above)
Comment (amcasey) [reply src/compiler/core.ts:1491]: Offline, we decided that an explanatory comment would suffice.

**Comment by amcasey** [comment]
Re-opening to see if that causes it to pick up the additional commits I pushed.

**Pull request merged**